### PR TITLE
Add test framework for Fourmolu options

### LINKED ISSUE #49: Testing all options
Until recently, all our output options have been binary choices*, with a neat split in to "the Ormolu version" and "the Fourmolu version". Our tests have thus tested our default options, and Ormolu's default options. This was never going to last. #48 is the first to introduce an option which is not used by default.

We have a few choices:
- Run all the tests with all the combinations of options. This may take a long time to run (I haven't tried it yet), and it would start to become difficult to review all the outputs to check that everything looks sensible. But it does give us pretty comprehensive regression tests.
  - Obviously this grows exponentially as we add more options...
- Only test each option on the examples where it should be relevant, e.g. there's no point running with and without `record-brace-space` on a file that doesn't contain any records. Setting this up would require reviewing each test case to determine which options are relevant. We could also miss bugs where an option affects something it shouldn't.
- Consider which options are likely to interact with each other, and filter the set of combinations accordingly. This has some of the same drawbacks as the previous choice.

\* Apart from indentation, but for the purposes of this discussion that's a choice between two and four. There doesn't seem much point testing with any more than that.

### LINKED ISSUE #56: Confusing docs for `diff-friendly-import-export`.
The docs for this option say "Whether to make use of extra commas in import/export lists (as opposed to Ormolu's style)", but Ormolu uses trailing commas on multi-line import and export lists (and I'm pretty sure it has since before fourmolu was forked), which would seem to be the "diff friendly" style.

In #19, tweag/ormolu#409 is referenced, but that issue isn't about commas, but rather the position of the parens.

So, I guess I'm asking for that documentation to be more explicit about how that option actually affects the formatting.

### PR DESCRIPTION
**Pull request by brandonchinn178**
Resolves #49, Resolves #56

- [x] Revert `PrinterSpec.hs` + `data/` to basically what it is in Ormolu
- [x] Add tests for current options
- [x] Add checklist to docs for adding a new config option
- [x] Add to docs list of current options

### CODE REVIEW

**Comment by brandonchinn178** [comment]
cc @georgefst Thoughts on both the test strategy + the input files? Did I cover the main styling differences for each option?

**Comment by arybczak** [comment]
Hey. Is this PR blocked by anything?

**Comment by brandonchinn178** [comment]
Hey @arybczak! I was hoping to get @georgefst to take a look, but he seems to be busy. I'm actually going on vacation, so I was going to merge this when I get back, give George a little more time anyway

**Comment by brandonchinn178** [comment]
FYI @3kyro this reverts the tests you added in #108, with new tests captured in https://github.com/fourmolu/fourmolu/pull/169/files#diff-2ce4cd9cc4e15f9f8fdd33721a2b90f5eade9582ca70bc09e6af9a9928fc5346. Can you check to see if there are any other cases we should test here?

@georgefst To get PRs unblocked and get activity flowing again in this repo, I'm going to merge this soon. It'd be great to get your review still, whether it comes before or after I merge.

File: data/fourmolu/import-export/output-Leading.hs [hunk 1]
```diff
@@ -1,3 +1,10 @@
+module Foo
+    ( foo
+    , bar
+    , baz
+    )
+where
+
 import qualified MegaModule as M
```
Comment (3kyro) [inline data/fourmolu/import-export/output-Leading.hs:8]: I think we should add a multi line import (or export) to test that commas respect the import-export list option (as per #187). It's probably best to check alignment of section headings and comments as well.
````haskell
module My.Module
    ( -- * Something
      foo
    , bar
      -- | A multiline
      -- comment here
    , bar2

      -- * Another thing
    , (<?>)
    )

import qualified MegaModule as M
    ( Either
    , Monad
        ( return
        , (>>)
        , (>>=)
        )
    , (<<<)
    , (>>>)
    )
````

**Comment by brandonchinn178** [comment]
@3kyro thanks! Updated here: https://github.com/fourmolu/fourmolu/pull/169/commits/1f80c2d148b76127c08f0a220daad5cf03f3cd57

The style's a bit wonky, but we can leave that off for later. I'll merge this in when CI passes

**Comment by 3kyro** [comment]
> The style's a bit wonky, but we can leave that off for later. I'll merge this in when CI passes

@brandonchinn178  Yes, styling with leading and diff-friendly is terrible. A ternary option would be great to have  but we can start by ignoring the diff-friendly setting when import export commas are set to leading. Let me know if there's anything I can do.

**Comment by brandonchinn178** [comment]
@3kyro ah yes ok, I did mention that here: https://github.com/fourmolu/fourmolu/pull/108#issuecomment-1100916957. Nothing to do now. Maybe we can take a look after I release the next version

**Pull request merged**